src/Sema.zig

```diff
@@ -23539,31 +23539,50 @@ fn zirErrorCast(sema: *Sema, block: *Block, extended: Zir.Inst.Extended.InstData
     // operand must be defined since it can be an invalid error value
     const maybe_operand_val = try sema.resolveDefinedValue(block, operand_src, operand);
 
+    // is the destination type the IES of the current function we're in
+    var is_dest_ret_ies: bool = false;
+
     const disjoint = disjoint: {
         // Try avoiding resolving inferred error sets if we can
         if (!dest_ty.isAnyError(zcu) and dest_ty.errorSetIsEmpty(zcu)) break :disjoint true;
         if (!operand_ty.isAnyError(zcu) and operand_ty.errorSetIsEmpty(zcu)) break :disjoint true;
         if (dest_ty.isAnyError(zcu)) break :disjoint false;
         if (operand_ty.isAnyError(zcu)) break :disjoint false;
-        const dest_err_names = dest_ty.errorSetNames(zcu);
-        for (0..dest_err_names.len) |dest_err_index| {
-            if (Type.errorSetHasFieldIp(ip, operand_ty.toIntern(), dest_err_names.get(ip)[dest_err_index]))
-                break :disjoint false;
-        }
 
-        if (!ip.isInferredErrorSetType(dest_ty.toIntern()) and
-            !ip.isInferredErrorSetType(operand_ty.toIntern()))
-        {
-            break :disjoint true;
-        }
+        // This will now be a resolved error set.
+        const resolved_dest_ty = if (ip.isInferredErrorSetType(dest_ty.toIntern())) blk: {
+            const dst_ies_func_index = ip.iesFuncIndex(dest_ty.toIntern());
+
+            if (sema.fn_ret_ty_ies) |dst_ies| {
+                if (dst_ies.func == dst_ies_func_index) {
+                    // dest_ty is the IES of the function we're currently in
+                    // we just append the operand's error set to our IES, thus making
+                    // them not disjoint.
+                    try dst_ies.addErrorSet(operand_ty, ip, sema.arena);
```

Comment:
Hm, I think this happens too late. For instance, the `isAnyError` case above may be hit. What does this code do?
```zig
fn foo() !void {
    const e: anyerror = error.Foo;
    const my_ies = @typeInfo(@typeInfo(foo).@"fn".return_type.?).error_union.error_set;
    const casted: my_ies = @errorCast(e);
    // i'm pretty sure that since `operand_ty.isAnyError(zcu)`, we break from the block early, and don't hit the `addErrorSet` call
    // so, the IES will be empty, and `casted` will be an invalid value!
    _ = casted;
}
```
I think this specific case needs to happen before all other checks. Probably just compute `is_dest_ret_ies` eagerly and handle it at the very top of `disjoint`.

Reply:
Great point, I've adjusted it as you suggested :)